Show itinerary results activity

akwc10 opened the pull request.

akwc10 committed:
Pass serialized domain model via Intent

Assigning views in activity.

akwc10 committed:
WIP Itinerary results recycler view with list adapter

Show network error.

akwc10 committed:
ItineraryResultsActivity and listAdapter clean up

WIP ItineraryResultsLegsActivity MVP boiler plating.

akwc10 committed:
WIP Convert itinerary results activity to fragment

akwc10 committed:
WIP Replace journeys fragment with legs fragment via callback

akwc10 committed:
Replaced journeys fragment with legs fragment via callback

Added corresponding layouts for fragments and associated recycler text views.
Added corresponding list adapters.
Cleanup.

akwc10 committed:
Removed androidx legacy support and updated material components

fsbarata commented on app/src/main/java/com/my/journeyplanner/views/itineraryresults/ItineraryResultsFragment.kt:
```diff
@@ -0,0 +1,81 @@
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        arguments?.let {
```
argument handling is usually done when view is created which is the callback onViewCreated

akwc10 replied:
Done

fsbarata replied:
onCreateView is different from onViewCreated. I suggest you have a look at the docs again for this

akwc10 replied:
Fixed

fsbarata commented on app/src/main/java/com/my/journeyplanner/views/itineraryresults/ItineraryResultsFragment.kt:
```diff
@@ -0,0 +1,81 @@
+        if (context is OnItineraryResultsFragmentInteractionListener) {
+            listener = context
+        } else {
+            throw RuntimeException("$context must implement OnItineraryResultsFragmentInteractionListener")
```
In kotlin, it is more common to throw other types of exception since there is no enforcement in the java checked exceptions. Thus I recommend a IllegalState here

akwc10 replied:
Done

fsbarata commented on app/src/main/java/com/my/journeyplanner/views/itineraryresults/ItineraryResultsFragment.kt:
```diff
@@ -0,0 +1,81 @@
+
+    override fun onDetach() {
+        super.onDetach()
+        listener = null
```
this is highly stylistic and not very relevant in Android:

usually teardown operations should be performed in the opposite order they were setup. In this case this would mean the super would be called last. Inconsequential though

akwc10 replied:
Done

fsbarata commented on app/src/main/java/com/my/journeyplanner/views/itineraryresults/ItineraryResultsListAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+        val view = holder.view
+        val context = view.context
+        val journey = getItem(position)
+        view.findViewById<TextView>(R.id.textViewStartTime).text = context.getString(
```
because BindViewHolder is called multiple times for one view holder, it is wasteful to keep finding the views. The best approach is to use vals for these views in the ViewHolder class so you only find them once during the onCreateViewHolder fase.

akwc10 replied:
Done

fsbarata replied:
can't see it, have you comitted?

akwc10 replied:
Fixed

fsbarata commented on app/src/main/java/com/my/journeyplanner/views/results/ResultsActivity.kt:
```diff
@@ -0,0 +1,76 @@
+
+    override fun showItineraryResultsFragment(journeys: List<Journey>) {
+        supportFragmentManager.beginTransaction()
+            .add(R.id.fragmentResults, ItineraryResultsFragment.newInstance(journeys)).commit()
```
favor replace, otherwise you can have some nasty bugs as you expand the functionality

akwc10 replied:
Done

fsbarata commented on app/src/main/res/layout/activity_results.xml:
```diff
@@ -74,15 +74,45 @@
+        android:layout_marginTop="16dp"
+        android:layout_marginEnd="16dp"
+        android:layout_marginBottom="16dp"
+        app:layout_constraintBottom_toTopOf="@+id/textViewResult"
```
watch out for these. Do not @+id in references, this will lead to bugs

akwc10 replied:
Whoops! Missed this one

fsbarata commented on app/src/main/res/layout/activity_results.xml:
```diff
@@ -74,15 +74,45 @@
         app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent" />
+
+    <fragment
```
this ought to be a FrameLayout that hosts fragments so you can swap them with the fragmentmanager like you did in the code

akwc10 replied:
Done

fsbarata commented on app/src/main/res/layout/recycler_text_views_itinerary_results.xml:
```diff
@@ -0,0 +1,39 @@
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:background="?android:attr/selectableItemBackground"
```
only available on Lollipop, use support library instead: ?selectableItemBackground

akwc10 replied:
Updated

fsbarata commented on core/src/main/java/com/my/core/domain/JourneyPlannerResultDomainModel.kt:
```diff
@@ -12,24 +13,24 @@ sealed class JourneyPlannerResultDomainModel {
 
         data class Journey(
             val arrivalDateTime: LocalDateTime,
             val duration: Int,
```
is the duration not inferred from the depart and arrival times?

akwc10 replied:
Removed. Now inferred:
        textViewDuration.text = context.getString(
            R.string.textView_duration,
            Duration.between(journey.startDateTime, journey.arrivalDateTime).toMinutes().toString()
        )

fsbarata commented on presenter/src/main/java/com/my/presenter/results/ResultsContract.kt:
```diff
@@ -0,0 +1,25 @@
+
+class ResultsContract {
+    interface View : BaseContract.View {
+        fun showMapActivity(mapItineraryResults: List<Any>)
```
Any? can't we do better than that?

fsbarata replied:
Is it just a placeholder for now?

akwc10 replied:
Yep, it's a temporary placeholder

akwc10 committed:
Feedback changes

Moved fragment argument handling from onCreate() to onCreateView().
Changed RuntimeException to IllegalStateException.
Updated onDetach() teardown order to be opposite of setup order.
onBindViewHolder() views assigned to vals optimisation.
Replaced supportFragmentManager.beginTransaction().add() with .replace().
Moved fragment inside FrameLayout so fragment swapping can be handled by supportFragmentManager.
Replaced ?android:attr/selectableItemBackground with ?selectableItemBackground as the former is only available in Lollipop.
Inferred duration from depart and arrival times.

akwc10 committed:
Corrected feedback corrections

Moved fragment arguments handling from onCreateView() to onViewCreated().
Moved ListAdapter find views from onBindViewHolder() to MyViewHolder() as the former is called multiple times per view holder.
Added debug build variant.

fsbarata commented on app/build.gradle:
```diff
@@ -17,9 +17,15 @@ android {
             minifyEnabled false
             proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
+            applicationIdSuffix ".debug"
+            debuggable true
```
debug is debuggable XD

fsbarata commented on app/src/main/res/layout/activity_results.xml:
```diff
@@ -74,15 +74,22 @@
-        app:layout_constraintTop_toBottomOf="@id/textViewItineraryResults" />
+        app:layout_constraintTop_toBottomOf="@id/textViewItineraryResults">
+
+        <fragment
```
we need to fix this, I think you didn't understand how to use fragments

fsbarata commented on app/src/main/res/layout/fragment_itinerary_results.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
don't need this layout

akwc10 committed:
Removed extra fragment in FrameLayout part 1

akwc10 committed:
Remove extra fragment from FrameLayout - part 2

akwc10 committed:
Remove debuggable flag for debug build variant

DOH!

akwc10 committed:
Make recyclerview the root view
